napari/layers/vectors/vectors.py

```diff
@@ -773,15 +797,120 @@ def _get_value(self, position):
+
         """
+        LOGGER.debug('Vectors._update_mesh')
         vertices, triangles = generate_vector_meshes(
             self.data[:, :, list(self._dims_displayed)],
             self.edge_width,
             self.length,
         )
         self._mesh_vertices = vertices
         self._mesh_triangles = triangles
-        self._displayed_stored = copy(self._dims_displayed)
+
+    def _get_slice(self, request: _VectorSliceRequest) -> _VectorSliceResponse:
+        """New method"""
+        LOGGER.debug('Vectors._get_slice : %s', request)
+
+        # expand indices to include out of slice as needed,
+        # and get alpha values
+        slice_indices = self._get_slice_indices(request)
+        indices, alphas = self._slice_data(slice_indices, request.dims_not_displayed, request.out_of_slice_display)
+
+        # assume the mesh is up to date at this point -
+        # all mesh generation should be triggered by the new method `_update_mesh`
+
+        disp = list(request.dims_displayed)
+
+        # if there is no data, set empty
+        if len(request.data) == 0 or len(indices) == 0:
+            view_faces = np.array([[0, 1, 2]])
+            view_data = np.empty((0, 2, 2))
+            view_indices = []
+            view_vertices = np.zeros((3, request.ndisplay))
+        # if more than 2D, reduce the data and mesh dimensions
+        elif request.ndim > 2:
+            # TODO: is this line necessary?
+            indices, alphas = self._slice_data(slice_indices, request.dims_not_displayed, request.out_of_slice_display)
+
+            view_indices = indices
+            view_alphas = alphas
+            view_data = request.data[np.ix_(indices, [0, 1], disp)]
+            keep_inds = np.repeat(2 * indices, 2)
+            keep_inds[1::2] = keep_inds[1::2] + 1
+            if request.ndisplay == 3:
+                keep_inds = np.concatenate(
+                    [
+                        keep_inds,
+                        len(request.mesh_triangles) // 2 + keep_inds,
+                    ],
+                    axis=0,
+                )
+            view_faces = request.mesh_triangles[keep_inds]
+            view_vertices = request.mesh_vertices
+        else:
+            view_faces = request.mesh_triangles
+            view_vertices = request.mesh_vertices
+            view_data = request.data[:, :, disp]
+            view_indices = np.arange(request.data.shape[0])
+            view_alphas = 1.0
+
+        thumbnail = self._make_thumbnail(view_data, view_indices)
+
+        # prep for vispy by translating [z,y,x]->[x,y,z]
+        vertices = view_vertices[:, ::-1]
+
+        face_color = self._view_face_color(view_indices, view_alphas, request.ndisplay, request.ndim, request.edge_color)
+
+        if self._ndisplay == 3 and self.ndim == 2:
+            vertices = np.pad(vertices, ((0, 0), (0, 1)), mode='constant')
+
+        return _VectorSliceResponse(
+            request=request,
+            data=request.data,
```

Comment:
The intent was for the response's data to be the subset of the request's data that is in the output slice. So maybe we should put `view_data` here instead?

Reply:
yes, I wasn't sure if I needed to keep `data` _an_ `view_data`. I agree with just keeping `view_data` though.

Reply:
ah, also, data is included in the base `_LayerSliceResponse` and `_LayerSliceRequest`. For vectors, the requests holds the original `layer.data`, but the response one needs to hold `view_data`. 

And if that's true, then I will keep the `view_` prefix for just the data and remove it from the rest (See discussion below)

Reply:
actually, since `data` is included in the base, I will need to include it in the `_VectorSliceResponse` as well (unless we move it out to only exist on the layers where its needed)

Reply:
Done.